### ISSUE TITLE
tag empty-heading rule with wcag2aa and wcag246

### ISSUE THREAD

**Issue by robfentress** [dequelabs/axe-core#700]
Shouldn't the `empty-heading` rule be tagged as failing **WCAG Success Criteria 2.4.6: Headings and Labels**?  I don't think it is simply "best practice."  

SC 2.4.6 states: 

> Headings and labels describe topic or purpose.

How could an empty heading ever describe what a section contains?

**Comment by WilcoFiers** [comment]
To the best of my knowledge, empty heading elements are ignored by assistive technologies. As I understand it, that means that empty heading elements aren't actually headings as far as WCAG 2 is concerned. So because empty heading elements aren't actually headings, the requirement of them being descriptive doesn't apply.

Even so, not everything mark up as a heading actually is a heading. There is no definition of headings in WCAG, but if you look at the definition of label you'll see that it has nothing to do with what markup was used for it. Something isn't a label because you used the label element. It's a label because it is presented as and functions as a label. It's a safe bet that that's true for headings as well. Empty headings are just garbage markup. You can fill a page with them and have it be totally accessible.

Hope that answers your question. Feel free to close the issue if you have no further comments.

**Comment by robfentress** [comment]
According to PowerMapper at least, [not all AT ignore empty headings](https://www.powermapper.com/tests/screen-readers/headings/h1-empty/).  Can you point me to the specification which clarifies how assistive technologies are _supposed_ to treat headings?  On the face of it, it seems ambiguous.

While it does appear that the [Understanding WCAG 2.4.6 page](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-descriptive.html#key-terms) explicitly states that a label, as used in this particular SC, is not synonymous with the element's accessible name, I'm not entirely convinced that this means that when a `<label>` element _is_ used, it does not have to describe it's referent.  Basically, while a "label," as used here, doesn't have to be a `<label>` element, all `<label>` elements _are_ labels and, thus, have to be descriptive.  Just so, headings.

Also, a Google search for "empty headings accessibility" returns, as its first result, a link to a Deque University page titled [Headings Must Not Be Empty](https://dequeuniversity.com/rules/axe/1.1/empty-heading).  This page makes reference to the rule id, empty-heading, which I assume is the same rule we're referring to.  On that page, it seems to indicate violating this rule fails [WCAG SC 1.3.1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html).  Thus, to be consistent with your documentation, it should at least be tagged with wcag2a and wcag131, if I'm correctly understanding your tagging system.

Additionally, that same Deque page also makes reference to [WCAG General Technique 130](https://www.w3.org/TR/WCAG20-TECHS/G130.html) (G130).  The G130 page on W3C, in turn, makes reference to WCAG SC 2.4.6.  Thus, transitively, one could make the case that the Deque page, itself, is supporting the interpretation that empty headings violate 2.4.6.  

Therefore, after further research, I continue to maintain that empty-heading should _not_ be tagged best-practice, but _should_ be tagged with wcag2a and wcag246, and, _additionally_, wcag131.

**Comment by robfentress** [comment]
Actually, looking at the way your tagging system works, it appears you may only tag something with wcag2a _or wcag2aa.  I assume that is because you consider things that fail wcag2a to automatically fail wcag2aa (?).  Thus, this oneshould be tagged wcga2a, wcag246, wcag131, and cat.name-role-value.  I guess?  Or am I misunderstanding how the tagging system works?

**Comment by WilcoFiers** [comment]
@robdodson The page you refer to is old aXe-core documentation. This is incorrect, and we did fix this as of the 2.4 release, but we haven't gone back to fix older pages. I've reached out to our docs writers to see if we can't exclude those older docs from getting drudged up by search engines.

Regarding all `<label>` elements being WCAG labels (and by extension all `<h#>` elements being WCAG headings). I don't think there is anything in the normative sections of WCAG 2 that says you're not allowed to have empty headings if they are ignored by assistive technologies.

Looking at the PowerMapper data, it looks like all the screen readers we list in our [baseline](https://github.com/dequelabs/axe-core/blob/develop/doc/accessibility-supported.md) will ignore empty headings. So that's no reason to flag it as a violation either. You can certainly configure aXe differently to better reflect your baseline, but the way it works currently is reflective of the capabilities of assistive technologies that we target by default.

**Comment by marcysutton** [comment]
I think you pinged the wrong Rob.

**Comment by WilcoFiers** [comment]
I think I did :P. Sorry Mr Dodson

**Comment by dylanb** [comment]
1. If you could assume 100% of the time that an empty heading means that there is some text missing from the page, then items found under this rule could and should be flagged as violations of WCAG under 2.4.6. Unfortunately, most of the time this is simply not the case.
2. From an accessibility supported perspective, the relevant screen readers ignore empty headings

Therefore, this is not a violation, it is a best practice. I suggest we close this issue.

@robfentress with respect to our approach to teaching, testing and developing:

1) When we test existing applications for conformance, we exclude best practices (unless requested otherwise)
2) When we teach people about accessibility, we teach them best practices
3) When developers are writing code, it is easy for them to implement best practices, therefore aXe turns best practice rules on because it is targeted primarily at developers

Therefore Deque University information will reference best practices

**Comment by WilcoFiers** [comment]
Closing the issue, since this seems to be more an interpretation preference, and aXe-core does have the option to change these preferences.
